### PR TITLE
Allow Kokkos::resize to take an execution space

### PR DESCRIPTION
**Pull request by masterleinad**
Related to #4792, #4823, and #4826. Specifying an execution space, we could at least get rid of a global fence (and possibly some instance fences depending on #4823 and #4826).
If the View's count is 1, we of course still deallocate data which implies a global fence.

~~So far tests are missing and~~ this is mostly intended to discuss if we want to provide these overloads.

### CODE REVIEW

**Comment by masterleinad** [comment]
Retest this please.

File: containers/src/Kokkos_DualView.hpp [hunk 1]
```diff
@@ -953,10 +953,29 @@ class DualView : public ViewTraits<DataType, Arg1Type, Arg2Type, Arg3Type> {
   ///
   /// This method only copies the old contents into the new View
   /// objects for the device which was last marked as modified.
-  template <class... I>
+  template <class... ViewCtorArgs>
   void impl_resize(const size_t n0, const size_t n1, const size_t n2,
                    const size_t n3, const size_t n4, const size_t n5,
-                   const size_t n6, const size_t n7, const I&... arg_prop) {
+                   const size_t n6, const size_t n7,
+                   const Impl::ViewCtorProp<ViewCtorArgs...>& arg_prop) {
```
Comment (ldh4) [inline containers/src/Kokkos_DualView.hpp:960]: Would it be too invasive of a change to make the param ordering of `impl_resize` to be consistent with `Kokkos::resize`?

Comment (masterleinad) [reply containers/src/Kokkos_DualView.hpp:960]: Not really, it was just the smallest change to the status quo.

Comment (masterleinad) [reply containers/src/Kokkos_DualView.hpp]: @ldh4 See f055b8c3e.

**Review by ldh4 — APPROVED** [review]

File: containers/src/Kokkos_DualView.hpp [hunk 2]
```diff
@@ -953,10 +953,29 @@ class DualView : public ViewTraits<DataType, Arg1Type, Arg2Type, Arg3Type> {
...
+        "The view constructor arguments passed to Kokkos::resize must "
+        "not include a pointer!");
+    static_assert(
+        !alloc_prop_input::has_memory_space,
+        "The view constructor arguments passed to Kokkos::resize must "
+        "not include a memory space instance!");
+    static_assert(
+        !alloc_prop_input::allow_padding,
+        "The view constructor arguments passed to Kokkos::resize must "
+        "not explicitly allow padding!");
```
Comment (crtrott) [inline containers/src/Kokkos_DualView.hpp:977]: Why this one?

File: core/src/Kokkos_CopyViews.hpp [hunk 3]
```diff
@@ -862,14 +862,16 @@ template <class DstType, class SrcType, class ExecSpace>
 struct ViewRemap<DstType, SrcType, ExecSpace, 1> {
   using p_type = Kokkos::pair<int64_t, int64_t>;
 
-  ViewRemap(const DstType& dst, const SrcType& src) {
+  template <typename... OptExecSpace>
+  ViewRemap(const DstType& dst, const SrcType& src,
+            const OptExecSpace&... exec_space) {
     if (dst.extent(0) == src.extent(0)) {
       view_copy(dst, src);
```
Comment (crtrott) [inline core/src/Kokkos_CopyViews.hpp:869]: why no exec_space forwarding here?

File: core/src/Kokkos_CopyViews.hpp [hunk 4]
```diff
@@ -878,7 +880,9 @@ template <class DstType, class SrcType, class ExecSpace>
 struct ViewRemap<DstType, SrcType, ExecSpace, 2> {
   using p_type = Kokkos::pair<int64_t, int64_t>;
 
-  ViewRemap(const DstType& dst, const SrcType& src) {
+  template <typename... OptExecSpace>
+  ViewRemap(const DstType& dst, const SrcType& src,
+            const OptExecSpace&... exec_space) {
     if (dst.extent(0) == src.extent(0)) {
       if (dst.extent(1) == src.extent(1)) {
         view_copy(dst, src);
```
Comment (crtrott) [inline core/src/Kokkos_CopyViews.hpp:888]: why no exec_space forwarding here?

File: core/src/Kokkos_CopyViews.hpp [hunk 5]
```diff
@@ -862,14 +862,16 @@ template <class DstType, class SrcType, class ExecSpace>
 struct ViewRemap<DstType, SrcType, ExecSpace, 1> {
   using p_type = Kokkos::pair<int64_t, int64_t>;
 
-  ViewRemap(const DstType& dst, const SrcType& src) {
+  template <typename... OptExecSpace>
+  ViewRemap(const DstType& dst, const SrcType& src,
+            const OptExecSpace&... exec_space) {
```
Comment (crtrott) [inline core/src/Kokkos_CopyViews.hpp:867]: what is exec_space... ?? If it includes more then execspace the name should probably be different?

File: core/src/Kokkos_CopyViews.hpp [hunk 6]
```diff
@@ -2872,19 +2915,33 @@ bool size_mismatch(const ViewType& view, unsigned int max_extent,
...
+                "must not include a label!");
+  static_assert(!alloc_prop_input::has_pointer,
+                "The view constructor arguments passed to Kokkos::resize must "
+                "not include a pointer!");
+  static_assert(!alloc_prop_input::has_memory_space,
+                "The view constructor arguments passed to Kokkos::resize must "
+                "not include a memory space instance!");
+  static_assert(!alloc_prop_input::allow_padding,
+                "The view constructor arguments passed to Kokkos::resize must "
+                "not explicitly allow padding!");
```
Comment (crtrott) [inline core/src/Kokkos_CopyViews.hpp:2944]: Why not padding?

**Review by crtrott — CHANGES_REQUESTED** [review]
A few questions.

File: containers/src/Kokkos_DualView.hpp (hunk 2 above)
Comment (masterleinad) [reply containers/src/Kokkos_DualView.hpp:977]: In the discussion around https://github.com/kokkos/kokkos/pull/5035#issuecomment-1131722266 we said that we shouldn't be allowed to explicitly request padding but I'm happy to change if that's what we want.

File: core/src/Kokkos_CopyViews.hpp (hunk 3 above)
Comment (masterleinad) [reply core/src/Kokkos_CopyViews.hpp]: Good catch! Fixed in 38b6001cb.

File: core/src/Kokkos_CopyViews.hpp (hunk 5 above)
Comment (masterleinad) [reply core/src/Kokkos_CopyViews.hpp:867]: It's just an execution space but it's optional to provide it.

**Comment by masterleinad** [comment]
Seems to work with `Sacado`.

**Review by crtrott — APPROVED** [review]

**Pull request merged**